src/lib/optimizer/chunk_statistics.hpp

```diff
@@ -0,0 +1,171 @@
+static constexpr uint32_t MAX_RANGES_COUNT = 10;
+
+class BaseFilter : public std::enable_shared_from_this<BaseFilter> {
+ public:
+  virtual ~BaseFilter() = default;
+
+  /**
+   * checks whether the filter is able to determine that the given value 
+   * and predicate condition will not yield any positive results with the values
+   * represented by the filter data.
+   * 
+   * In other words: A scan operation with value and predicate_type on the column/chunk
+   * that this filter was created on would yield zero result rows.
+  */
+  virtual bool can_prune(const AllTypeVariant& value, const PredicateCondition predicate_type) const = 0;
+};
+
+class ChunkColumnStatistics {
+ public:
+  static std::shared_ptr<ChunkColumnStatistics> build_statistics(DataType data_type,
+                                                                 std::shared_ptr<BaseColumn> column);
+
+  void add_filter(std::shared_ptr<BaseFilter> filter) { _filters.emplace_back(filter); }
+
+  bool can_prune(const AllTypeVariant& value, const PredicateCondition predicate_type) const {
+    for (const auto& filter : _filters) {
+      if (filter->can_prune(value, predicate_type)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+ protected:
+  std::vector<std::shared_ptr<BaseFilter>> _filters;
+};
+
+template <typename T>
+class MinMaxFilter : public BaseFilter {
+ public:
+  explicit MinMaxFilter(T min, T max) : _min(min), _max(max) {}
+  ~MinMaxFilter() override = default;
+
+  bool can_prune(const AllTypeVariant& value, const PredicateCondition predicate_type) const override {
+    const auto t_value = type_cast<T>(value);
+    // Operators work as follows: value_from_table <operator> t_value
+    // e.g. OpGreaterThan: value_from_table > t_value
+    // thus we can exclude chunk if t_value >= _max since then no value from the table can be greater than t_value
+    switch (predicate_type) {
+      case PredicateCondition::GreaterThan:
+        return t_value >= _max;
+      case PredicateCondition::GreaterThanEquals:
+        return t_value > _max;
+      case PredicateCondition::LessThan:
+        return t_value <= _min;
+      case PredicateCondition::LessThanEquals:
+        return t_value < _min;
+      case PredicateCondition::Equals:
+        return t_value < _min || t_value > _max;
+      default:
+        return false;
+    }
+  }
+
+ protected:
+  const T _min;
+  const T _max;
+};
+
+template <typename T>
+class RangeFilter : public BaseFilter {
+ public:
+  static_assert(std::is_arithmetic_v<T>, "RangeFilter should not be instantiated for strings.");
+
+  explicit RangeFilter(std::vector<std::pair<T, T>> ranges) : _ranges(std::move(ranges)) {}
+  ~RangeFilter() override = default;
+
+  static std::unique_ptr<RangeFilter<T>> build_filter(const pmr_vector<T>& dictionary);
+
+  bool can_prune(const AllTypeVariant& value, const PredicateCondition predicate_type) const override {
+    const auto t_value = type_cast<T>(value);
```

Comment:
ugh, type_cast. You know the concrete type (T) and thus should use boost::get<T>(value), don't you?

Reply:
`type_cast` uses `boost::get` internally.